src/main/kotlin/com/github/avrokotlin/avro4k/encoder/AbstractAvroEncoder.kt

```diff
@@ -0,0 +1,217 @@
+package com.github.avrokotlin.avro4k.encoder
+
+import com.github.avrokotlin.avro4k.AvroConfiguration
+import com.github.avrokotlin.avro4k.getAvroFullName
+import com.github.avrokotlin.avro4k.isListOfBytes
+import com.github.avrokotlin.avro4k.schema.getTypeNamed
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.SerializationStrategy
+import kotlinx.serialization.descriptors.PolymorphicKind
+import kotlinx.serialization.descriptors.PrimitiveKind
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.SerialKind
+import kotlinx.serialization.descriptors.StructureKind
+import kotlinx.serialization.encoding.CompositeEncoder
+import kotlinx.serialization.encoding.Encoder
+import org.apache.avro.AvroRuntimeException
+import org.apache.avro.Schema
+import org.apache.avro.generic.GenericData
+import org.apache.avro.generic.GenericData.EnumSymbol
+import org.apache.avro.generic.GenericFixed
+import java.math.BigDecimal
+import java.nio.ByteBuffer
+
+
+@OptIn(ExperimentalSerializationApi::class)
+abstract class AbstractAvroEncoder : Encoder, NativeAvroEncoder {
```

Comment:
Note here, we are using Encoder instead of AbstractEncoder, to split only value encoding (GenericAvroEncoder) and structure encoding (map, list and record). AbstractEncoder implements both Encoder and CompositeEncoder